File: AdvancedSortingAlgorithm/CyclicSort/findDuplicatesPart-2.py
from typing import List

class Solution:
    def findDuplicates(self, nums: List[int]) -> List[int]:
        n = len(nums)
        i = 0
        while i < n:
            correctedIdx = nums[i]-1
            if nums[i] == i+1 or nums[i] == nums[correctedIdx]:
                i += 1
            else:
                nums[i],nums[correctedIdx] = nums[correctedIdx],nums[i]
        ans = []
        for i in range(n):
            if nums[i] != i+1:
                ans.append(nums[i])
        return ans
        

File: AdvancedSortingAlgorithm/CyclicSort/findDuplicatesPart-2.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> findDuplicates(vector<int>& nums) {
        int n=nums.size();
        int i=0;
        vector<int>ans;
        while(i<n){
            int correctedIdx = nums[i]-1;
            if(nums[i]==i+1 || nums[correctedIdx]==nums[i]){
                i+=1;
            }
            else{
                swap(nums[i],nums[correctedIdx]);
            }
        }
        for(int i=0;i<n;i++){
            if(nums[i]!=i+1){
                ans.push_back(nums[i]);
            }
        }
        return ans;
    }
};